### PR TITLE
Refactor of backend API

### PR DESCRIPTION
**Pull request by lvella**
Would like feedback first before changing eStark.

### CODE REVIEW

File: pipeline/src/pipeline.rs [hunk 1]
```diff
@@ -724,15 +724,16 @@ impl<T: FieldElement> Pipeline<T> {
     }
 
     fn maybe_write_proof(&self, proof_result: &ProofResult<T>) -> Result<(), Vec<String>> {
-        if let Some(constraints_serialization) = &proof_result.constraints_serialization {
+        // TODO: this must go inside the backend...
```
Comment (leonardoalt) [inline pipeline/src/pipeline.rs:727]: I disagree with this. IMO the user needs to access the artifact and do whatever it needs with it

File: cli/src/main.rs [hunk 2]
```diff
@@ -624,32 +624,23 @@ fn verification_key<T: FieldElement>(
         .with_setup_file(params.map(PathBuf::from))
         .with_backend(*backend_type);
 
-    let vkey = pipeline.verification_key()?;
-
-    write_verification_key_to_fs(vkey, dir);
+    let mut vkey_file = BufWriter::new(fs::File::create(dir.join("vkey.bin")).unwrap());
+    pipeline.write_verification_key(&mut vkey_file)?;
+    vkey_file.flush().unwrap();
+    log::info!("Wrote vkey.bin.");
```
Comment (leonardoalt) [inline cli/src/main.rs:630]: I find it odd that here the pipeline writes to FS, and in the setup case it returns the data and the user writes to FS.

File: pipeline/src/pipeline.rs [hunk 3]
```diff
@@ -865,7 +866,10 @@ impl<T: FieldElement> Pipeline<T> {
         self.arguments.query_callback.as_deref()
     }
 
-    pub fn verification_key(&mut self) -> Result<Vec<u8>, Vec<String>> {
+    pub fn write_verification_key<W: io::Write>(
```
Comment (leonardoalt) [inline pipeline/src/pipeline.rs:869]: Why does Pipeline write it to FS instead of just returning the key?

File: backend/src/lib.rs [hunk 4]
```diff
@@ -169,12 +58,14 @@ pub enum Error {
...
     #[error("the backend has no setup operations")]
     NoSetupAvailable,
-    #[error("proof generation failed")]
-    ProofFailed(String),
-    #[error("proof verification failed")]
-    VerificationFailed(String),
-    #[error("verification key generation failed")]
-    VerificationKeyFailed(String),
+    #[error("backend specific error")]
+    BackendError(String),
```
Comment (leonardoalt) [inline backend/src/lib.rs:62]: I think we could keep the old errors since they are generic enough for the things we want to do.

File: backend/src/lib.rs [hunk 5]
```diff
@@ -184,91 +75,40 @@ pub type Proof = Vec<u8>;
...
-        fixed: &[(String, Vec<F>)],
-    ) -> Result<Vec<u8>, Error>;
+        pil: &'a Analyzed<F>,
+        fixed: &'a [(String, Vec<F>)],
+        setup: Option<&mut dyn io::Read>,
+        verification_key: Option<&mut dyn io::Read>,
+    ) -> Result<Box<dyn Backend<'a, F> + 'a>, Error>;
+
+    /// Generate a new setup.
+    fn generate_setup(&self, size: DegreeType, output: &mut dyn io::Write) -> Result<(), Error>;
```
Comment (leonardoalt) [inline backend/src/lib.rs:90]: Why is setup the only function that goes into this trait?

Comment (leonardoalt) [reply backend/src/lib.rs:90]: shouldn't it also go into the Backend trait like the other functionalities?

Comment (lvella) [reply backend/src/lib.rs:90]: This is because how we use it in `powdr setup` (I think), where we can call it without needing the analyzed pil or the fixed columns. Verifying and proving, on the other hand, do require the full object with the common stuff.

File: pipeline/src/pipeline.rs (hunk 3 above)
Comment (lvella) [reply pipeline/src/pipeline.rs:862]: Why is the pipeline returning a key in the first place?

I am very uncomfortable with uninterpreted `Vec<u8>` being passed around contained backend specific serialized data. We are bypassing the type system. This is no better than casting to `void *` or using `unsafe` to reinterpret cast. Serializations should go directly to disk.

File: backend/src/lib.rs (hunk 4 above)
Comment (lvella) [reply backend/src/lib.rs:62]: I changed it because this way it is simpler and less error prone to write the backend implementation: you can  just use `?` to turn return from a `Result<_, String>` and it becomes a backend error. Whether it is a proving or a verification error, it is obvious from the function that was called.

File: cli/src/main.rs (hunk 2 above)
Comment (leonardoalt) [reply cli/src/main.rs:629]: Because at the moment of writing that code the pipeline was the easiest way to collect all required artifacts at the same time. Try collecting the verification key in a different way and you'll see. You'll need to repeatedly call the Pipeline to get different artifacts, or call the specific crates directly.

Comment (leonardoalt) [reply cli/src/main.rs:629]: That said, I'm fine with this not being in Pipeline and if it's easy to remove it and do it like Setup is done, sounds good to me.

File: pipeline/src/pipeline.rs (hunk 3 above)
Comment (leonardoalt) [reply pipeline/src/pipeline.rs:862]: I disagree it should go directly to disk. I think the user should decide what to do with it. Vec<u8> is a universal serialized data type.

File: backend/src/lib.rs (hunk 4 above)
Comment (leonardoalt) [reply backend/src/lib.rs:62]: Sure, not too strongly opinionated here.

File: pipeline/src/pipeline.rs (hunk 3 above)
Comment (lvella) [reply pipeline/src/pipeline.rs:862]: The proposed API accommodates both `File` and `Vec<u8>`. Both implement `io::Write`.

File: backend/src/lib.rs [hunk 6]
```diff
@@ -184,91 +75,40 @@ pub type Proof = Vec<u8>;
...
-where
-    Self: Sized + BackendImpl<F>,
-{
-    /// Create a backend object from a setup loaded from a file.
-    fn new_from_setup(input: &mut dyn io::Read) -> Result<Self, io::Error>;
 
-    /// Write the setup to a file.
-    fn write_setup(&self, output: &mut dyn io::Write) -> Result<(), io::Error>;
+    /// Write the verification key.
+    fn write_verification_key(&self, output: &mut dyn io::Write) -> Result<(), Error>;
```
[the remark this answers is not in the document]
Comment (leonardoalt) [reply backend/src/lib.rs:113]: Vec<u8> is universal enough. But sure, fine with `serialize`

File: backend/src/lib.rs (hunk 5 above)
Comment (leonardoalt) [reply backend/src/lib.rs:90]: I see. Ok... I still find it ugly that a backend that has nothing to do with setup needs to actively implement this function and refuse it, but well.. that's the trade off of having the dynamic interfaces

Comment (lvella) [reply backend/src/lib.rs:90]: Oh, yeah. I forgot about it, but I was planning to provide default implementation to everything that is optional for a backend to have.

File: backend/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (leonardoalt) [reply backend/src/lib.rs]: Actually, fine with "write"

File: backend/src/lib.rs (hunk 5 above)
Comment (lvella) [reply backend/src/lib.rs]: Much better now.

File: backend/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (lvella) [reply backend/src/lib.rs]: Now it is called "export_".

File: pipeline/src/verify.rs [hunk 7]
```diff
@@ -11,7 +11,7 @@ pub fn verify(temp_dir: &Path, name: &str, constants_name: Option<&str>) {
         temp_dir.to_str().unwrap()
     );
     let commits_file = format!("{}/{name}_commits.bin", temp_dir.to_str().unwrap());
-    let constraints_file = format!("{}/{name}_constraints.json", temp_dir.to_str().unwrap());
+    let constraints_file = format!("{}/constraints.json", temp_dir.to_str().unwrap());
```
Comment (leonardoalt) [inline pipeline/src/verify.rs:14]: why?

File: pipeline/src/pipeline.rs [hunk 8]
```diff
@@ -681,10 +682,9 @@ impl<T: FieldElement> Pipeline<T> {
 
     fn maybe_write_constants(&self, constants: &[(String, Vec<T>)]) -> Result<(), Vec<String>> {
         if let Some(path) = self.path_if_should_write(|name| format!("{name}_constants.bin"))? {
-            write_polys_file(
-                &mut BufWriter::new(&mut fs::File::create(path).unwrap()),
-                constants,
-            );
+            let mut file = BufWriter::new(fs::File::create(path).unwrap());
+            write_polys_file(&mut file, constants);
+            file.flush().unwrap();
```
Comment (leonardoalt) [inline pipeline/src/pipeline.rs:687]: Why not keep it as a separate function? These 3 lines are duplicated in other places

File: pipeline/src/pipeline.rs [hunk 9]
```diff
@@ -874,16 +864,28 @@ impl<T: FieldElement> Pipeline<T> {
...
+                let mut setup_file = self
+                    .arguments
+                    .setup_file
+                    .as_ref()
+                    .map(|path| BufReader::new(fs::File::open(path).unwrap()));
+
+                let backend = factory
+                    .create(
+                        pil.borrow(),
+                        &fixed_cols[..],
```
Comment (leonardoalt) [inline pipeline/src/pipeline.rs:877]: In general I find it annoying to pass the fixed_cols around like this (which is not this PR's fault). We should have some type "low level program" that has the PIL constraints AND the fixed columns.

File: pipeline/src/pipeline.rs [hunk 10]
```diff
@@ -51,8 +51,6 @@ pub struct ProofResult<T: FieldElement> {
     pub witness: Option<Vec<(String, Vec<T>)>>,
     /// Proof, potentially None (if success is false)
     pub proof: Option<Proof>,
-    /// Serialized low level constraints, potentially None (if success is false)
-    pub constraints_serialization: Option<String>,
```
Comment (leonardoalt) [inline pipeline/src/pipeline.rs:55]: why?

Comment (leonardoalt) [reply pipeline/src/pipeline.rs:55]: don't we need it for the tests?

File: cli/src/main.rs [hunk 11]
```diff
@@ -624,32 +623,23 @@ fn verification_key<T: FieldElement>(
         .with_setup_file(params.map(PathBuf::from))
         .with_backend(*backend_type);
 
-    let vkey = pipeline.verification_key()?;
-
-    write_verification_key_to_fs(vkey, dir);
+    let mut vkey_file = BufWriter::new(fs::File::create(dir.join("vkey.bin")).unwrap());
+    pipeline.export_verification_key(&mut vkey_file)?;
+    vkey_file.flush().unwrap();
```
Comment (leonardoalt) [inline cli/src/main.rs:628]: a bit ugly to always have to do this flush

File: backend/src/lib.rs [hunk 12]
```diff
@@ -184,91 +72,51 @@ pub type Proof = Vec<u8>;
...
 
-/// Trait implemented by backends that have a setup phase that must be saved to
-/// a file.
-trait BackendImplWithSetup<F: FieldElement>
-where
-    Self: Sized + BackendImpl<F>,
-{
-    /// Create a backend object from a setup loaded from a file.
-    fn new_from_setup(input: &mut dyn io::Read) -> Result<Self, io::Error>;
+    /// Exports the setup in a backend specific format. Can be used to create
```
Comment (leonardoalt) [inline backend/src/lib.rs:111]: wdym "can be used to create a new instance of the same backend"?

File: halo2/src/prover.rs [hunk 13]
```diff
@@ -36,90 +35,89 @@ use std::{
...
+pub use halo2_proofs::poly::commitment::Params;
 pub use halo2_proofs::poly::kzg::commitment::ParamsKZG;
-
-/// Create a halo2 proof for a given PIL, fixed column values and witness column values
-/// We use KZG ([GWC variant](https://eprint.iacr.org/2019/953)) and Keccak256
-pub struct Halo2Prover {
+pub use halo2_proofs::SerdeFormat;
+
+/// Create a halo2 proof for a given PIL, fixed column values and witness column
+/// values We use KZG ([GWC variant](https://eprint.iacr.org/2019/953)) and
```
Comment (leonardoalt) [inline halo2/src/prover.rs:43]:
```suggestion
/// values. We use KZG ([GWC variant](https://eprint.iacr.org/2019/953)) and
```

File: halo2/src/prover.rs [hunk 14]
```diff
@@ -132,27 +130,16 @@ impl Halo2Prover {
...
     ) -> Result<Vec<u8>, String> {
         log::info!("Starting proof aggregation...");
 
-        // TODO this is hacky
-        let degree = usize::BITS - pil.degree().leading_zeros() + 1;
-        let params_app = {
-            let mut params = self.params.clone();
-            params.downsize(degree);
-            params
-        };
```
Comment (leonardoalt) [inline halo2/src/prover.rs:150]: Why was this removed? This is important because the universal setup may be a common large one, and here we want to downsize it to the circuit's degree

File: halo2/src/prover.rs [hunk 15]
```diff
@@ -219,32 +206,19 @@ impl Halo2Prover {
...
-        pil: &Analyzed<F>,
-        fixed: &[(String, Vec<F>)],
-    ) -> Result<VerifyingKey<G1Affine>, String> {
-        let plaf_circuit = analyzed_to_plaf(pil, fixed);
-        let circuit = analyzed_to_circuit_with_zeroed_witness(plaf_circuit, pil);
-        match keygen_vk(&self.params, &circuit) {
-            Ok(vkey) => Ok(vkey),
-            Err(e) => Err(e.to_string()),
-        }
+    pub fn add_verification_key(&mut self, mut vkey: &mut dyn io::Read) {
```
Comment (leonardoalt) [inline halo2/src/prover.rs:209]: I guess we don't need this function anymore?

File: halo2/src/prover.rs [hunk 16]
```diff
@@ -219,32 +206,19 @@ impl Halo2Prover {
...
     }
 
-    pub fn verification_key<F: FieldElement>(
-        &self,
-        pil: &Analyzed<F>,
-        fixed: &[(String, Vec<F>)],
-    ) -> Result<Vec<u8>, String> {
-        let vk = self.verification_key_inner(pil, fixed)?;
-        Ok(vk.to_bytes(SerdeFormat::Processed))
+    pub fn verification_key(&self) -> Result<VerifyingKey<G1Affine>, String> {
```
Comment (leonardoalt) [inline halo2/src/prover.rs:219]: Maybe this could already check if we have one and simply clone that?

**Review by leonardoalt — COMMENTED** [review]
Nice! Some small comments:

File: pipeline/src/verify.rs (hunk 7 above)
Comment (lvella) [reply pipeline/src/verify.rs:14]: Would be another argument into the backend. Maybe the backend should take instead a reference to the pipeline.

File: pipeline/src/pipeline.rs (hunk 10 above)
Comment (lvella) [reply pipeline/src/pipeline.rs:55]: I hope not. I remember running all the tests last friday. This was just so we export the json to the prover-cpp.

File: cli/src/main.rs (hunk 11 above)
Comment (lvella) [reply cli/src/main.rs:628]: This is just so write errors aren't ignored. I thought of two solutions for this:

1. a new BufWriter-like struct that panics if there is an error flushing inside `drop()` (instead of ignoring it);
2. a plain function that encapsulates this repeated code.

File: backend/src/lib.rs (hunk 12 above)
Comment (lvella) [reply backend/src/lib.rs:111]: "can be used to create a new instance of the same backend type" makes it clearer? Means that I can pass the exported setup to `create()` function.

File: halo2/src/prover.rs (hunk 15 above)
Comment (lvella) [reply halo2/src/prover.rs:209]: As I said previously, I think `powdr_halo2` belongs as a mod inside backend crate, but I didn't want to do any any deeper change to the `Halo2Prover` in this PR.

**Comment by leonardoalt** [comment]
test failing

File: halo2/src/prover.rs (hunk 14 above)
Comment (lvella) [reply halo2/src/prover.rs:150]: I have placed this `downsize()` in `create()`. Do you think it is correct there? Still fails the test.

Comment (leonardoalt) [reply halo2/src/prover.rs:150]: which test?

File: backend/src/lib.rs (hunk 12 above)
Comment (lvella) [reply backend/src/lib.rs]: reworded

File: halo2/src/prover.rs (hunk 14 above)
Comment (lvella) [reply halo2/src/prover.rs:150]: The one that was failing. But it is fixed now.

File: pipeline/src/pipeline.rs (hunk 8 above)
Comment (lvella) [reply pipeline/src/pipeline.rs:706]: Placed it in a different function.

File: cli/src/main.rs (hunk 11 above)
Comment (lvella) [reply cli/src/main.rs:628]: Wrote a new function for these calls.

File: halo2/src/prover.rs (hunk 16 above)
Comment (lvella) [reply halo2/src/prover.rs:226]: This mimics the behavior of what is in main. `verification_key()` always creates a new key, and if this key is not manually added to the backend, it panics when verifying.

Doesn't look good. Maybe if a key is needed, either by this function or by `verify()`, a new one should be created and set. Or either, create the key if none is supplied in `create()`.

Should I change this in this PR?

File: pipeline/src/pipeline.rs (hunk 10 above)
Comment (leonardoalt) [reply pipeline/src/pipeline.rs:55]: Yea it's tricky. I agree Pipeline shouldn't really need to know about this, but at the same time I also don't think a backend should be writing specific things to the disk.

File: halo2/src/prover.rs (hunk 16 above)
Comment (leonardoalt) [reply halo2/src/prover.rs:226]: The key is deterministic so I don't think we need to create a new one. But either is fine (here or another PR)

**Comment by leonardoalt** [comment]
please squash then I'll approve

File: halo2/src/prover.rs [hunk 17]
```diff
@@ -36,90 +35,96 @@ use std::{
...
-    ) {
+        setup: Option<&mut dyn io::Read>,
+    ) -> Result<Self, io::Error> {
+        Self::assert_field_is_bn254();
+
+        let params = setup
+            .map(|mut setup| ParamsKZG::<Bn256>::read(&mut setup))
+            .transpose()?
+            .map(|mut params| {
+                params.downsize(degree_bits(pil.degree()));
```
Comment (leonardoalt) [inline halo2/src/prover.rs:76]: I'm also not sure it should be done here. Question: if I want to compute two separate proofs, then aggregate them, do you see that as 3 backend instances, or would I use the same? If the same, this here isn't ideal because I might want to keep a large `setup` that works for all circuits, but when making specific proofs I downsize the setup params to that specific circuit.

Comment (lvella) [reply halo2/src/prover.rs:76]: If the PIL and fixed cols do change in between the separated proofs, this API is not very good. The degree is taken from the PIL, so it would require one instance per proof.

Comment (leonardoalt) [reply halo2/src/prover.rs:76]: they don't change for separate proofs, that's the whole point.

Comment (lvella) [reply halo2/src/prover.rs:76]: Then I don't understand. Either PIL and degree needs to change in between proofs, or they don't. The way it is, if the setup is loaded, it is downsized. If it is generated from the factory, it is not.

**Review by leonardoalt — APPROVED** [review]
